proposals/function-delegate.md

```diff
@@ -0,0 +1,195 @@
+
+### Function wrapper implementation
+
+Sometimes we need to add functionalities to function that needs to hold state. For example, when we need to add buffering to function. Let's say we have fibbonacci function:
+
+```
+fun fib(i: Int) = if (i <= 2) 1 else fib(i - 1) + fib(i - 1)
+```
+
+Without buffering it is very unefficient. Using function delegation we might add buffering this way:
+
+```
+fun <V, T> memoize(f: (V) -> T): (V) -> T {
+    val map = mutableMapOf<V, T>()
+    return { map.getOrPut(it) { f(it) } }
+}
+
+fun fib(i: Int): Int by memoize { i -> if (i <= 2) 1 else fib(i - 1) + fib(i - 1) }
+```
+
+Similarly synchronization for single function could be added.
+
+## Implementation details
+
+Function delegate could be compiled to delegate holden in seperate property and its invocation in function body. Invocation should include all the paramters defined by function. For example, above `fib` function with `memoize` should be compiled to:
+
+```
+private val fib$delegate: (Int) -> Int = memoize { i -> if (i <= 2) 1 else fib(i - 1) + fib(i - 1) }
+fun fib(i: Int) = fib$delegate(i)
+```
+
+similarly `setTitle` will be compiled:
+
+```
+class MainActivity: MainView {
+
+    private val setTitle#delegate = titleView::setText
+    override fun setTitle(text: String) = setTitle#delegate(text)
+}
+```
+
+## Alternatives
+
+If we are using property instead of function when we can get all the functionalities. For example:
+
+```
+val fib: (Int) -> Int = memoize<Int, Int> { a ->
+    if (a <= 2) 1
+    else (::fib).get()(a - 1) + (::fib).get()(a - 2)
+}
+```
+
+The problem with this alternative is when we are using recurrence (because property is not defined yet). Also there is probem with interfaces, because we would have to replace functions with properties with function types:
+
+```
+interface MainView {
+    val clearList: ()->Unit
+}
+
+class ListAdapter {
+    fun clear() {}
+}
+
+class MainActivity : MainView {
+    val adapter = ListAdapter()
+    
+    override val clearList: ()->Unit = adapter::clear
+}
+```
+
+## Other notation proposition
+
+Instead of `as` keyword, we could use simpler notation:
+
+```
+fun fib = memoize<Int, Int> { a -> if (a <= 2) 1 else fib(a - 1) + fib(a - 2) }
+```
+
+Type of arguments would be inferred from delegate. Argument types would be inferred from delegate.
+
+## Open questions
+
+Is `by` keyword a good choice while it is reminding property delegation, and as oposed to it, here are are not passing context and function reference? What if one day we would want to make function delegate that actually acts like property delegate:
+
+```
+class Delegate {
+    operator fun functionDelegate(thisRef: Any?, function: KFunction<*>, val property: Int): Int // Designed for this function
+    operator fun functionDelegate(thisRef: Any?, function: KFunction<*>, vararg arguments: Any?): Int // Desifned for any functio
```

Comment:
This is a better and more general solution -- there's no reason that the rest of the above proposal couldn't be implemented using such operator functions on `KFunction[X]`. I feel like the proposal can just be reduced to support of something like `operator fun functionDelegate(thisRef: Any?, function: KFunction<*>, vararg arguments: Any?)`.

Reply:
Fact is that would be more coherent with property delegation, and I can also make extension functions that would allow previous use-cases:
```
fun (()->Unit).functionDelegate(thisRef: Any?, function: KFunction<*>) = this()

fun <T1> ((T1)->Unit).functionDelegate(thisRef: Any?, function: KFunction<*>, arg1: T1) = this(arg1)

fun <T1, T2> ((T1, T2)->Unit).functionDelegate(thisRef: Any?, function: KFunction<*>, arg1: T1, arg2: T2) = this(arg1, arg2)
```
But I would l would like to hear whar others are thinking in this subject.